sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java

```diff
@@ -110,35 +91,42 @@ public static Optional<ServiceInstance> chooseServiceInstanceByXds(String servic
         }
         List<ServiceInstance> serviceInstanceList = new ArrayList<>(serviceInstanceByXdsRoute);
         XdsLoadBalancer loadBalancer = XdsLoadBalancerFactory
-                .getLoadBalancer(serviceInstanceList.get(0).getClusterName());
+                .getLoadBalancer(serviceName, serviceInstanceList.get(0).getClusterName());
         return Optional.of(loadBalancer.selectInstance(serviceInstanceList));
     }
 
     /**
      * isXdsRouteRequired
      *
-     * @param host host
+     * @param serviceName serviceName
      * @return isXdsRouteRequired
      */
-    public static boolean isXdsRouteRequired(String host) {
-        // if host is ip, so no xds routing required
-        if (StringUtils.isEmpty(host) || host.equals(LOCAL_HOST) || IP_PATTERN.matcher(host).matches()) {
+    public static boolean isXdsRouteRequired(String serviceName) {
+        // if service is localhost or started not with lowercase, so no xds routing required
+        if (StringUtils.isEmpty(serviceName) || serviceName.equals(LOCAL_HOST)
+                || isNotStartWithLowercase(serviceName)) {
             return false;
         }
         return true;
     }
 
+    private static boolean isNotStartWithLowercase(String serviceName) {
+        char firstChar = serviceName.charAt(0);
+        return firstChar < 'a' || firstChar > 'z';
+    }
+
     /**
      * process headers, just get first value for every header
      *
      * @param headers headers
      * @return processed headers
      */
     public static Map<String, String> processHeaders(Map<String, List<String>> headers) {
-        return headers.entrySet().stream()
-                .collect(Collectors.toMap(
-                        Entry::getKey,
-                        entry -> CollectionUtils.isEmpty(entry.getValue()) ? "" : entry.getValue().get(0)
-                ));
+        Map<String, String> processedHeaders = new HashMap<>();
+        for (Entry<String, List<String>> headerEntry : headers.entrySet()) {
+            processedHeaders.put(headerEntry.getKey(),
+                    CollectionUtils.isEmpty(headerEntry.getValue()) ? "" : headerEntry.getValue().get(0));
+        }
+        return processedHeaders;
```

Comment:
headerEntry.getValue() is probably null

Reply:
if headerEntry.getValue() is null or empty, put "" value into the map

Reply:
done